src/dynamic_graph_head/controllers.py

```diff
@@ -29,18 +30,119 @@ def __init__(self, head, Kp, Kd, with_sliders=False):
+        self.joint_positions = head.get_sensor("joint_positions")
+        self.joint_velocities = head.get_sensor("joint_velocities")
 
         if with_sliders:
-            self.slider_positions = head.get_sensor('slider_positions')
+            self.slider_positions = head.get_sensor("slider_positions")
 
     def warmup(self, thread_head):
         self.zero_pos = self.joint_positions.copy()
 
         if self.with_sliders:
             self.slider_zero_pos = self.map_sliders(self.slider_positions)
 
+    def go_zero(self):
+        # TODO: Make this an interpolation.
+        self.zero_pos = np.zeros_like(self.zero_pos)
+
+        if self.with_sliders:
+            self.slider_zero_pos = self.map_sliders(self.slider_positions)
+
+    def map_sliders(self, sliders):
+        sliders_out = np.zeros_like(self.joint_positions)
+
+        if self.joint_positions.shape[0] == 12:  # solo12
+            slider_A = sliders[0]
+            slider_B = sliders[1]
+            for i in range(4):
+                sliders_out[3 * i + 0] = slider_A
+                sliders_out[3 * i + 1] = slider_B
+                sliders_out[3 * i + 2] = 2.0 * (1.0 - slider_B)
+
+                if i >= 2:
+                    sliders_out[3 * i + 1] *= -1
+                    sliders_out[3 * i + 2] *= -1
+
+            # Swap the hip direction.
+            sliders_out[3] *= -1
+            sliders_out[9] *= -1
+
+        elif self.joint_positions.shape[0] == 8:  # solo8
+            slider_A = sliders[0]
+            for i in range(4):
+                sliders_out[2 * i + 0] = slider_A
+                sliders_out[2 * i + 1] = 2.0 * (1.0 - slider_A)
+
+                if i >= 2:
+                    sliders_out[2 * i + 0] *= -1
+                    sliders_out[2 * i + 1] *= -1
+
+        elif self.joint_positions.shape[0] == 6:  # bolt
+            slider_A = sliders[0]
+            slider_B = sliders[1]
+
+            for i in range(2):
+                sliders_out[3 * i + 0] = slider_A
+                sliders_out[3 * i + 1] = slider_B
+                sliders_out[3 * i + 2] = 1.0 - slider_B
+
+            sliders_out[3] *= -1
+
+        elif self.joint_positions.shape[0] == 2:  # teststand
+            slider_A = sliders[0]
+            sliders_out[0] = slider_A
+            sliders_out[1] = 2.0 * (1.0 - slider_A)
+
+        return sliders_out
+
+    def run(self, thread_head):
+        if self.with_sliders:
+            self.des_position = (
+                self.slider_scale
+                * (self.map_sliders(self.slider_positions) - self.slider_zero_pos)
+                + self.zero_pos
+            )
+        else:
+            self.des_position = self.zero_pos
+
+        self.tau = (
+            self.Kp * (self.des_position - self.joint_positions)
+            - self.Kd * self.joint_velocities
+        )
+        self.head.set_control("ctrl_joint_torques", self.tau)
+
+
+class HoldOnBoardPDController:
```

Comment:
I wonder if you could make `HoldOnBoardPDController` a subclass from `HoldPDController`? This way the `map_sliders` etc. method would need to be implemented only once and reused across the two classes.

What do you think?

Reply:
Good point

Reply:
Not yet tested, so please don't merge yet!